[Flow EVM] Add support for `ABI` encoding/decoding of Solidity `byte` types

m-Peter opened the pull request:
Closes: https://github.com/onflow/flow-go/issues/5365

Support `ABI` encoding/decoding the following Solidity `byte` types:
- `bytes` & `bytes[]`
- `bytes4` & `bytes4[]`
- `bytes32` & `bytes32[]`

**Note:** A contract upgrade for `EVM` system contract is needed.

The test payloads have been generated with the help of: https://abi.hashex.org/

turbolent commented:
Great work adding support for more types!

I had a look at the documentation at https://docs.soliditylang.org/en/latest/types.html#fixed-size-byte-arrays, and from what I understood, the Solidity `bytes` type (dynamic-sized byte array) is equivalent to the Cadence type `[UInt8]` (variable-sized array), the `bytesN` type (fixed-size byte-array) is equivalent to the Cadence type `[UInt8; N]` (constant-sized array).

Why do we need the "wrapper types", like `EVM.EVMBytes4`?

m-Peter commented:
> Great work adding support for more types!
> 
> I had a look at the documentation at https://docs.soliditylang.org/en/latest/types.html#fixed-size-byte-arrays, and from what I understood, the Solidity `bytes` type (dynamic-sized byte array) is equivalent to the Cadence type `[UInt8]` (variable-sized array), the `bytesN` type (fixed-size byte-array) is equivalent to the Cadence type `[UInt8; N]` (constant-sized array).
> 
> Why do we need the "wrapper types", like `EVM.EVMBytes4`?

@turbolent Let me explain this with a working example, to illustrate the differences:
```javascript
let bytes4 = '0x050a0f14' // these are the values 5, 10, 15, 20 in hex
let abiEncoded = web3.eth.abi.encodeParameter('bytes4', bytes4)
assert.equal(
    abiEncoded,
    '0x050a0f1400000000000000000000000000000000000000000000000000000000'
)
    
let uint8four = [5, 10, 15, 20] // this is the same value as `bytes4` variable above
abiEncoded = web3.eth.abi.encodeParameter('uint8[4]', uint8four)
assert.equal(
    abiEncoded,
    '0x0000000000000000000000000000000000000000000000000000000000000005000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000f0000000000000000000000000000000000000000000000000000000000000014'
)

let bytes = '0x050a0f14'
abiEncoded = web3.eth.abi.encodeParameter('bytes', bytes)
assert.equal(
    abiEncoded,
    '0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000004050a0f1400000000000000000000000000000000000000000000000000000000'
)
```

The Solidity `bytes`-related types, are probably used for compact encoding, compared to Solidity arrays.

Right now, we have mapped the Cadence dynamic-sized & fixed-sized arrays, to Solidity arrays:
- [UInt8] -> uint8
- [UInt8; 4] -> uint8[4]

Which means that we need to somehow control the ABI encoding/decoding of Solidity `bytes`-related types. That's why I've introduced the wrapper types.
- `EVM.EVMBytes` -> `bytes`
- `EVM.EVMBytes4` -> `bytes4`
- `EVM.EVMBytes32` -> `bytes32`

These are the most commonly used Solidity `bytes`-related types. I have also added support for arrays of those types. Let me know if this makes sense :pray:

turbolent commented:
Thank you for the explanation, it makes sense now! Initially I assumed that the byte types were just aliases, but I understand now that they are independent types, to allow for more compact encoding.

Ideally we would probably like to have a type like `EVM.EVMBytes<N>`, which contains a `[UInt8; N]`, and where `N` is a compile-time constant. However, Cadence does not support generic types in Cadence code yet.

The code looks good, and adding support for the common types makes sense. Maybe we can consider generating the code for all `N` `1...32` with `go:generate`, but we can do that later, if needed

turbolent commented:
As the EVM contract change, the state commitments need to be updated as well. You can just use the new reported values as the expected values.

m-Peter commented on fvm/evm/impl/abi.go:
```diff
@@ -307,16 +364,26 @@ func gethABIType(staticType interpreter.StaticType, evmAddressTypeID common.Type
 	case *interpreter.CompositeStaticType:
-		if staticType.TypeID != evmAddressTypeID {
-			break
+		if staticType.TypeID == evmTypeIDs.AddressTypeID {
```
Good point :+1: Updated in https://github.com/onflow/flow-go/pull/6553/commits/853543c9952abb4adfcf3fc41bd94fba6ccb555c

m-Peter commented:
> As the EVM contract change, the state commitments need to be updated as well. You can just use the new reported values as the expected values.

Updated in https://github.com/onflow/flow-go/pull/6553/commits/f067110dbcc8fcfec0715c43dec762d981efce63

bluesign commented:
> Ideally we would probably like to have a type like EVM.EVMBytes<N>, which contains a [UInt8; N], and where N is a compile-time constant. However, Cadence does not support generic types in Cadence code yet.

maybe we can do something like `EVM.createBytes<N>([UInt8;N]): EVM.EVMBytes<N>`  ?

turbolent commented:
@bluesign 

> Out of curiosity, can we do something like `EVM.createBytes<N>([UInt8;N]): EVM.EVMBytes<N>` and `EVM.createBytes([UInt8]): EVM.EVMBytes` ?

That would be nice, and the ideal solution. However, Cadence does not support generic types defined in Cadence programs yet, and further, it only supports generic parameters which are types at the moment, there is no support for generic parameters like numeric constants yet (see e.g. ["const generics" in Rust](https://blog.rust-lang.org/2021/03/25/Rust-1.51.0.html#const-generics-mvp))

turbolent reviewed (approved):
Great work! :clap:

janezpodhostnik reviewed (approved):
Nice!
